### PR TITLE
feat(Auction: SignerTimeout ): finish contract script and test

### PR DESCRIPTION
**Pull request by cosinlink**

### CODE REVIEW

**Review by huwenchao — DISMISSED** [review]
lgtm

File: contracts/toCKB-typescript/src/utils/config.rs [hunk 1]
```diff
@@ -1,8 +1,28 @@
+pub const UDT_LEN: usize = 16;
 pub const PLEDGE: u64 = 10000;
+pub const XT_CELL_CAPACITY: u64 = 100;
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/utils/config.rs:3]:
```suggestion
pub const XT_CELL_CAPACITY: u64 = 200;
```

File: contracts/toCKB-typescript/src/switch/auction_signertimeout/mod.rs [hunk 2]
```diff
@@ -1,7 +1,218 @@
...
+    // cap of toCKB_cell ==  XT_CELL_CAPACITY + to_bidder + to_trigger + to_signer
+    let collateral = load_cell_capacity(0, Source::GroupInput)? - XT_CELL_CAPACITY;
+    let mut to_bidder = collateral;
+    let init_collateral = collateral * AUCTION_INIT_PERCENT as u64 / 100;
+    if auction_time == 0 {
+        to_bidder = init_collateral
+    } else if auction_time < AUCTION_MAX_TIME {
+        to_bidder =
+            init_collateral + (collateral - init_collateral) * auction_time / AUCTION_MAX_TIME
+    }
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/auction_signertimeout/mod.rs:130]:
```suggestion
    if auction_time < AUCTION_MAX_TIME {
        to_bidder =
            init_collateral + (collateral - init_collateral) * auction_time / AUCTION_MAX_TIME
    }
```

File: contracts/toCKB-typescript/src/switch/auction_signertimeout/mod.rs [hunk 3]
```diff
@@ -1,7 +1,218 @@
...
+    Ok(())
+}
+
+fn verify_outputs(
+    input_data: &ToCKBCellDataView,
+    auction_time: u64,
+    toCKB_lock_hash: &[u8],
+    lot_amount: u128,
+) -> Result<(), Error> {
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/auction_signertimeout/mod.rs:104]: add comments to describe the fixed order of output cells

File: contracts/toCKB-typescript/src/switch/auction_signertimeout/mod.rs [hunk 4]
```diff
@@ -1,7 +1,218 @@
...
+    // check if it is timestamp value
+    {
+        return Err(Error::InputSinceInvalid);
+    }
+
+    let auction_time = since & VALUE_MASK;
+    Ok(auction_time)
+}
+
+fn verify_inputs(toCKB_lock_hash: &[u8], lot_amount: u128) -> Result<(), Error> {
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/auction_signertimeout/mod.rs:70]: add comments to describe the fixed order of input cells

File: tests/src/toCKB_typescript/tests/auction_signertimeout.rs [hunk 5]
```diff
@@ -1 +1,243 @@
...
+
+    let err = context.verify_tx(&tx, MAX_CYCLES).unwrap_err();
+    assert_error_eq!(
+        err,
+        ScriptError::ValidationFailure(Error::InputSinceInvalid as i8)
+    );
+}
+
+#[test]
+fn test_wrong_XT() {
```
Comment (mkxbl) [inline tests/src/toCKB_typescript/tests/auction_signertimeout.rs:64]: wrong_XT Error contains 2 kinds:
- wrong XT amount
- wrong XT cell capacity

better to test both

File: contracts/toCKB-typescript/src/switch/auction_signertimeout/mod.rs (hunk 2 above)
Comment (cosinlink) [reply contracts/toCKB-typescript/src/switch/auction_signertimeout/mod.rs:130]: if auction_time == 0, does to_bidder == collateral ?. In my opinion, it should == init_collateral that is 67% * collateral

File: contracts/toCKB-typescript/src/switch/auction_signertimeout/mod.rs [hunk 6]
```diff
@@ -1,7 +1,210 @@
...
+    let input_data = toCKB_data_tuple
+        .0
+        .as_ref()
+        .expect("inputs should contain toCKB cell");
+    let toCKB_lock_hash = load_cell_lock_hash(0, Source::GroupInput)?;
+
+    let lot_amount: u128 = match get_xchain_kind()? {
+        XChainKind::Btc => {
+            let btc_lot_size = input_data.get_btc_lot_size()?;
+            btc_lot_size.get_sudt_amount()
```
Comment (huwenchao) [inline contracts/toCKB-typescript/src/switch/auction_signertimeout/mod.rs:32]: It's better to impl a `get_sudt_amount` function in `ToCKBCellDataView`.

File: contracts/toCKB-typescript/src/switch/auction_signertimeout/mod.rs [hunk 7]
```diff
@@ -1,7 +1,210 @@
...
+    debug!("1. check bidder lock success! ");
+
+    // expect paying ckb to bidder,trigger and signer
+    // cap of toCKB_cell ==  XT_CELL_CAPACITY + to_bidder + to_trigger + to_signer
+    let asset_collateral = load_cell_capacity(0, Source::GroupInput)? - XT_CELL_CAPACITY;
+    let mut to_bidder = asset_collateral;
+    let init_collateral = asset_collateral * AUCTION_INIT_PERCENT as u64 / 100;
+    if auction_time < AUCTION_MAX_TIME {
+        to_bidder =
+            init_collateral + (asset_collateral - init_collateral) / AUCTION_MAX_TIME * auction_time
```
Comment (huwenchao) [inline contracts/toCKB-typescript/src/switch/auction_signertimeout/mod.rs:120]:
```suggestion
            init_collateral + (asset_collateral - init_collateral) * auction_time / AUCTION_MAX_TIME
```

we should always use `*` first to remain precision as much as possible.

**Pull request merged**